lib/services/access.go

```diff
@@ -59,3 +62,24 @@ type Access interface {
 	// ReplaceRemoteLocks replaces the set of locks associated with a remote cluster.
 	ReplaceRemoteLocks(ctx context.Context, clusterName string, locks []types.Lock) error
 }
+
+// CheckDynamicLabelsInDenyRules checks if any deny rules in the given role use
+// labels prefixed with "dynamic/".
+func CheckDynamicLabelsInDenyRules(r types.Role) error {
```

Comment:
That's quite tricky to do while parsing/before evaluating the expression, which is what we'd want. Label expressions use the [typical](https://github.com/gravitational/teleport/blob/047ecdfd60a165dbabb4aca3c020f27a0220c930/lib/services/label_expressions.go#L59) parser library which tries to be general-purpose and assumes everything including map keys could be subexpressions. So it would be very difficult to add some special case during the parse to check the allowed keys for this case. Especially while label expression support the [labels_matching](https://github.com/gravitational/teleport/blob/047ecdfd60a165dbabb4aca3c020f27a0220c930/lib/services/label_expressions.go#L115) function that can reference all labels where the key matches a regex

But, we could filter those labels out or return an error when the expression is actually being evaluated, pretty easy to add here https://github.com/gravitational/teleport/blob/047ecdfd60a165dbabb4aca3c020f27a0220c930/lib/services/label_expressions.go#L67. Unfortunately the user wouldn't see the error until they are actually trying to use the role.

You could try evaluating the expression with some dummy input to catch the error, but in general the evaluation could return an error for any number of reasons, even if the expression is perfectly valid, and I would prefer to avoid that

Reply:
There's also the option of doing a dumb regex search of the expression string for `"dynamic/`. I think it would be pretty unlikely to get false positives